RecoJets/JetProducers/src/MVAJetPuId.cc

```diff
@@ -0,0 +1,568 @@
+		tmvaVariables_.push_back( "fRing1" );
+		tmvaVariables_.push_back( "fRing2"   );
+		tmvaVariables_.push_back( "fRing3"  );
+		tmvaVariables_.push_back( "ptD"  );
+		tmvaVariables_.push_back( "beta"   );
+		tmvaVariables_.push_back( "betaStar"  );
+		tmvaVariables_.push_back( "DR_weighted"  );
+		tmvaVariables_.push_back( "pull" );
+		tmvaVariables_.push_back( "jetR");
+		tmvaVariables_.push_back( "jetRchg");
+
+		tmvaNames_["rho"] = "rho";
+		tmvaNames_["nTot"] = "nTot";
+		tmvaNames_["nCh"] = "nCh";
+		tmvaNames_["axisMajor"] = "axisMajor";
+		tmvaNames_["axisMinor"] = "axisMinor";
+		tmvaNames_["fRing0"] = "fRing0";
+		tmvaNames_["fRing1"] = "fRing1";
+		tmvaNames_["fRing2"] = "fRing2";
+		tmvaNames_["fRing3"] = "fRing3";
+		tmvaNames_["ptD"] = "ptD";
+		tmvaNames_["beta"] = "beta";
+		tmvaNames_["betaStar"] = "betaStar";
+		tmvaNames_["DR_weighted"] = "DR_weighted";
+		tmvaNames_["pull"] = "pull";  
+		tmvaNames_["jetR"] = "jetR";  
+		tmvaNames_["jetRchg"] = "jetRchg";
+}
+
+
+
+MVAJetPuId::~MVAJetPuId() 
+{
+	if( reader_ ) {
+		delete reader_;
+	}
+}
+
+
+void Assign(const std::vector<float> & vec, float & a, float & b, float & c, float & d )
+{
+	size_t sz = vec.size();
+	a = ( sz > 0 ? vec[0] : 0. );
+	b = ( sz > 1 ? vec[1] : 0. );
+	c = ( sz > 2 ? vec[2] : 0. );
+	d = ( sz > 3 ? vec[3] : 0. );
+}
+
+
+void SetPtEtaPhi(const reco::Candidate & p, float & pt, float & eta, float &phi )
+{
+	pt  = p.pt();
+	eta = p.eta();
+	phi = p.phi();
+}
+
+
+void MVAJetPuId::bookReader()
+{
+	reader_ = new TMVA::Reader("!Color:Silent");
+	assert( ! tmvaMethod_.empty() && !  tmvaWeights_.empty() );
+	for(std::vector<std::string>::iterator it=tmvaVariables_.begin(); it!=tmvaVariables_.end(); ++it) {
+		if(  tmvaNames_[*it].empty() ) { 
+			tmvaNames_[*it] = *it;
+		}
+		reader_->AddVariable( *it, variables_[ tmvaNames_[*it] ].first );
+	}
+	for(std::vector<std::string>::iterator it=tmvaSpectators_.begin(); it!=tmvaSpectators_.end(); ++it) {
+		if(  tmvaNames_[*it].empty() ) { 
+			tmvaNames_[*it] = *it;
+		}
+		reader_->AddSpectator( *it, variables_[ tmvaNames_[*it] ].first );
+	}
+	reco::details::loadTMVAWeights(reader_,  tmvaMethod_.c_str(), tmvaWeights_.c_str() ); 
+}
+
+
+void MVAJetPuId::set(const PileupJetIdentifier & id)
+{
+	internalId_ = id;
+}
+
+
+void MVAJetPuId::runMva()
+{
+	if( ! reader_ ) { bookReader();}
+	if(fabs(internalId_.jetEta_) <  5.0) internalId_.mva_ = reader_->EvaluateMVA( tmvaMethod_.c_str() );
+	if(fabs(internalId_.jetEta_) >= 5.0) internalId_.mva_ = -2.;
+	internalId_.idFlag_ = computeIDflag(internalId_.mva_,internalId_.jetPt_,internalId_.jetEta_);
+}
+
+std::pair<int,int> MVAJetPuId::getJetIdKey(float jetPt, float jetEta)
+{
+	int ptId = 0;                                                                                                                                    
+	if(jetPt > 10 && jetPt < 20) ptId = 1;                                                                                 
```

Comment:
the ranges should be semi-inclusive (jetPt with 20, and 30 has to get some ptId != 0)